lib/services/local/sessiontracker.go

```diff
@@ -0,0 +1,379 @@
+			session.AddParticipant(*update.AddParticipant.Participant)
+		case *proto.UpdateSessionTrackerRequest_RemoveParticipant:
+			session.RemoveParticipant(update.RemoveParticipant.ParticipantID)
+		}
+	default:
+		return trace.BadParameter("unrecognized session version %T", session)
+	}
+
+	sessionJSON, err := marshalSession(session)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	item := backend.Item{Key: backend.Key(sessionPrefix, req.SessionID), Value: sessionJSON}
+	_, err = s.bk.CompareAndSwap(ctx, *sessionItem, item)
+	if trace.IsCompareFailed(err) {
+		time.Sleep(retryDelay)
+		return s.UpdateSessionTracker(ctx, req)
+	}
+
+	return trace.Wrap(err)
+}
+
+// RemoveSessionTracker removes a tracker resource for an active session.
+func (s *sessionTracker) RemoveSessionTracker(ctx context.Context, sessionID string) error {
+	err := s.removeSessionFromList(ctx, sessionID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(s.bk.Delete(ctx, backend.Key(sessionPrefix, sessionID)))
+}
+
+func (s *sessionTracker) addSessionToList(ctx context.Context, sessionID string) error {
+	listItem, err := s.bk.Get(ctx, backend.Key(sessionPrefix, sessionList))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var list []string
+	err = utils.FastUnmarshal(listItem.Value, &list)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	list = append(list, sessionID)
+	listJSON, err := utils.FastMarshal(list)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newListItem := backend.Item{Key: backend.Key(sessionPrefix, sessionList), Value: listJSON}
+	_, err = s.bk.CompareAndSwap(ctx, *listItem, newListItem)
+	return trace.Wrap(err)
+}
+
+func (s *sessionTracker) removeSessionFromList(ctx context.Context, sessionID string) error {
+	listItem, err := s.bk.Get(ctx, backend.Key(sessionPrefix, sessionList))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var list []string
+	err = utils.FastUnmarshal(listItem.Value, &list)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	found := false
+	for i, id := range list {
+		session, err := s.loadSession(ctx, id)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		doGC := session.GetCreated().Add(gcDelay).Before(time.Now().UTC()) && session.GetState() == types.SessionState_SessionStateTerminated
+		isStale := session.GetExpires().Before(time.Now().UTC())
+		if id == sessionID || doGC || isStale {
+			list = append(list[:i], list[i+1:]...)
+			found = true
+			break
+		}
+
+		if doGC {
+			err := s.RemoveSessionTracker(ctx, id)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
+	if !found {
+		return trace.NotFound("session %v not found in list", sessionID)
+	}
+
+	listJSON, err := utils.FastMarshal(list)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	newListItem := backend.Item{Key: backend.Key(sessionPrefix, sessionList), Value: listJSON}
+	_, err = s.bk.CompareAndSwap(ctx, *listItem, newListItem)
+	if trace.IsCompareFailed(err) {
+		time.Sleep(retryDelay)
```

Comment:
We never want to do hard sleeps.  Use a select statement to allow for context cancellation.